src/ZODB/tests/racetest.py

```diff
@@ -0,0 +1,515 @@
+                    db.close()
+
+            try:
+                for i in range(N):
+                    # print('T%s.%d' % (tx, i))
+                    if failed.is_set():
+                        break
+                    t_()
+            except:  # noqa: E722 do not use bare 'except'
+                failed.set()
+                raise
+
+        # run the workers concurrently.
+        init()
+
+        N = 100 // (2*4)  # N reduced to save time
+        tg = []
+        for x in range(nwork):
+            t = threading.Thread(name='T%d' % x, target=T, args=(x, N))
+            t.start()
+            tg.append(t)
+
+        for t in tg:
+            t.join(60)
+
+        if failed.is_set():
+            self.fail('\n\n'.join([_ for _ in failure if _]))
+
+
+# _state_init initializes the database according to the spec.
+def _state_init(db, spec):
+    transaction.begin()
+    zconn = db.open()
+    root = zconn.root()
+    spec.init(root)
+    spec.assertStateOK(root)
+    transaction.commit()
+    zconn.close()
+
+
+# _state_invalidate_half1 invalidatates first 50% of database objects, so that
+# the next time they are accessed, they are reloaded from the storage.
+def _state_invalidate_half1(root):
+    keys = list(sorted(root.keys()))
+    for k in keys[:len(keys)//2]:
+        obj = root[k]
+        obj._p_invalidate()
+
+
+# _state_details returns text details about ZODB objects directly referenced by
+# root.
+def _state_details(root):  # -> txt
+    # serial for all objects
+    keys = list(sorted(root.keys()))
+    txt = ''
+    txt += '  '.join('%s._p_serial: %s' % (k, tid_repr(root[k]._p_serial))
+                     for k in keys)
+    txt += '\n'
+
+    # zconn.at approximated as max(serials)
+    # XXX better retrieve real zconn.at, but currently there is no way to
+    # retrieve it for all kind of storages.
+    zconn = root._p_jar
+    zconn_at = max(root[k]._p_serial for k in keys)
+    txt += 'zconn_at: %s  # approximated as max(serials)\n' % \
+           tid_repr(zconn_at)
+
+    # zstor.loadBefore(obj, @zconn_at)
+    zstor = zconn.db().storage
+
+    def load(key):
+        load_txt = 'zstor.loadBefore(%s, @zconn.at)\t->  ' % key
+        obj = root[key]
+        x = zstor.loadBefore(obj._p_oid, at2before(zconn_at))
+        if x is None:
+            load_txt += 'None'
+        else:
+            _, serial, next_serial = x
+            load_txt += 'serial: %s  next_serial: %s' % (
+                        tid_repr(serial), tid_repr(next_serial))
+        load_txt += '\n'
+        return load_txt
+
+    for k in keys:
+        txt += load(k)
+
+    # try to reset storage cache and retry loading
+    # it helps to see if an error was due to the cache getting out of sync
+    zcache = getattr(zstor, '_cache', None)  # works for ZEO and NEO
+    if zcache is not None:
+        zcache.clear()
+        txt += 'zstor._cache.clear()\n'
+        for k in keys:
+            txt += load(k)
+
+    return txt
```

Comment:
Could you put this functionality into a model base class?

The model could then look like:
```
@implementer(IModelSpec)
class Model(object):
    def __init__(self, db): ...
    def initialize(self): ... initialize the database (your init)...
    def modify(self): ... modifiy the database (your next)...
    def verify(self): ... verify the database (your assertStateOk + logic to provide failure details) ...
```

Reply:
Like I've tried to explain in https://github.com/zopefoundation/ZODB/pull/368#discussion_r959707435 my idea is to keep models minimal. Models just define ways for database state to be evolved and to assert that intended invariant is held. Everything else - to verify model behaviour on particular usage scenario - is the job of the model checker.

In specific domains sometimes it is possible to implement such checker to exercise all possible behaviors. Here I'm talking about formal methods that verify systems.

In our practical case, it is not possible to exercise all possible behaviors, because there are so many states of the system, that verifying them all takes time close to infinity. That's why there are particular checkers, that try to exercise particular scenarios with the intention to catch bugs a particular checker was designed for.

But still the approach is to have a definition for _all_ possible evolution of the system (via "init" and "next" steps), and then to verify all reachable states wrt the invariant.

From this point of view models should be minimal and whole the mechanics for the verification should reside in the checker(s). We might indeed try to unify the code in between various checkers one way or another. The way it is done currently via factoring common functionality - that checkers all use - into common functions might be not perfect, but, I believe, it should be ok in practice. Still we can discuss how to amend this provided that we don't move checking functionality into the models.

P.S. The way verification is done here is inspired from what I've learned from [TLA+](https://lamport.azurewebsites.net/tla/tla.html). I believe [TLA+ Video Course](https://lamport.azurewebsites.net/video/videos.html) is especially useful, because, even if/when one does not use TLA+ to verify/specify system, it still teaches one an approach about how to think more clearly about concurrency and distributed systems.